Add initial neo datetime FFI

sffc opened the pull request:
#5940

This is the DateTime version. I plan to do basically the exact same design for:

- Time
- Date
- GregorianDate
- GregorianDateTime
- ZonedDateTime
- GregorianZonedDateTime

See datetime.cpp for usage examples.

This supports both a builder API and a comprehensive static field set API by using the `with_fset` function I added.

It is all in the FFI crate, which means this is less likely to conflict with some of the other open PRs in the datetime crate (CC @robertbastian).

robertbastian reviewed (changes requested):
Issue: please replace the existing API instead of adding an additional API. Any `Neo` name that you introduce now will just need to be cleaned up next week, and any code you don't delete today will have to be maintained until someone gets around to deleting it (we were in this state in the datetime crate for too long, it sucked).

Also, I would like to review a diff, not a brand new API that I manually have to compare to the old one. This applies to both the Diplomat definitions, as well as the C++ and Dart tests.

Manishearth commented:
I think the point here is to migrate one API, and then if that goes well migrate the rest with the same pattern, but the APIs are interconnected so it's not possible to do that without "Neo". I guess an alternate route would be to have this PR as a draft to _show_ the new API, get it reviewed but not landed, and then do the whole migration.

sffc commented:
Like in Rust, the new FFI API is a different shape than the old FFI API, and I plan to delete (rather than migrate) the old one, which I am happy to do in the same PR. I am asking you to review the shape of the new API to catch any architectural issues before I spend time copying it into all the rest of the types.

sffc commented:
I marked it as draft to indicate that I don't plan to land this in its current form, but I would still like a review of it.

sffc commented:
Old C++:

```cpp
    std::unique_ptr<DateTimeFormatter> any_dtf = DateTimeFormatter::create_with_length(*locale.get(), DateTimeLength::Medium).ok().value();
    out = any_dtf->format(*any_date.get(), *any_time.get()).ok().value();
    std::cout << "Formatted value is " << out << std::endl;
    if (out != "Oct 5, 2 Reiwa, 1:33\u202fPM") {
        std::cout << "Output does not match expected output" << std::endl;
        return 1;
    }
```

New C++, static field set:

```cpp
    std::unique_ptr<NeoDateTimeFormatter> fmt_ymdt = NeoDateTimeFormatter::create_ymdt(*locale.get(), NeoDateTimeLength::Medium, TimePrecision::Minute, DateTimeAlignment::Auto, YearStyle::Auto).ok().value();
    out = fmt_ymdt->format_iso(*date.get(), *time.get());
    std::cout << "Fieldset YMDT: " << out;
    if (out != "11 jul 2022, 13:06") {
        std::cout << " (unexpected!)";
        saw_unexpected_output = true;
    }
    std::cout << std::endl;
```

New C++, dynamic field set:

```cpp
    DateTimeFieldSetBuilder builder;
    builder.length = std::optional<NeoDateTimeLength>(NeoDateTimeLength::Long);
    builder.date_fields = std::optional<DateFields>(DateFields::YM);
    std::unique_ptr<NeoDateTimeFormatter> fmt_ym_bld = NeoDateTimeFormatter::create_from_builder(*locale.get(), builder).ok().value();
    out = fmt_ym_bld->format_iso(*date.get(), *time.get());
    std::cout << "Fieldset YM in DateTimeFormatter via builder: " << out;
    if (out != "julio de 2022") {
        std::cout << " (unexpected!)";
        saw_unexpected_output = true;
    }
    std::cout << std::endl;
```

(these examples are not doing the exact same thing but they are all end-to-end formatting call sites)

sffc commented:
Some specific things I want comments on:

Field sets are modeled as constructors with positional arguments. For example, in Rust there is a struct with three fields:

https://unicode-org.github.io/icu4x/rustdoc/icu/datetime/fieldsets/struct.MDT.html

In C++, I modeled it as a constructor with the same three positional arguments.

I do not currently have a discriminator in the constructor name as we do in other places. Should I add one? For example, if we add a new semantic option in the future (not unlikely), it would need to be added as a positional argument. We have the `_mv1` as an escape hatch, so I don't consider this a hard requirement.

I could also use a struct. I would name the struct based on what it contains, not what it does, so that the same struct can be shared across multiple field set constructors.

Or, I could use the builder struct in the field set constructors, and make the constructor fail at runtime if the builder struct is not compatible with the requested field set.

For example:

```cpp
// Currently in PR:
// Infallible except for data-related errors:
NeoDateTimeFormatter::create_ymdt(
    *locale.get(),
    NeoDateTimeLength::Medium,
    TimePrecision::Minute,
    DateTimeAlignment::Auto,
    YearStyle::Auto
)

// Alternative A:
// Infallible except for data-related errors:
NeoDateTimeFormatter::create_ymdt_with_length_time_precision_alignment_and_year_style(
    *locale.get(),
    NeoDateTimeLength::Medium,
    TimePrecision::Minute,
    DateTimeAlignment::Auto,
    YearStyle::Auto
)

// Alternative B:
LengthTimePrecisionAlignmentYearStyleOptions options;
options.length = std::optional<NeoDateTimeLength>(NeoDateTimeLength::Medium);
options.time_precision = std::optional<TimePrecision>(TimePrecision::Minute);
options.alignment = std::optional<DateTimeAlignment>(DateTimeAlignment::Auto);
options.year_style = std::optional<YearStyle>(YearStyle::Auto);
// Infallible except for data-related errors:
NeoDateTimeFormatter::create_ymdt_with_options_v1(
    *locale.get(),
    options
)

// Alternative C:
DateTimeFieldSetBuilder builder;
builder.length = std::optional<NeoDateTimeLength>(NeoDateTimeLength::Medium);
builder.time_precision = std::optional<TimePrecision>(TimePrecision::Minute);
builder.alignment = std::optional<DateTimeAlignment>(DateTimeAlignment::Auto);
builder.year_style = std::optional<YearStyle>(YearStyle::Auto);
// Fails if the builder contains options incompatible with the field set, as well as data-related errors:
NeoDateTimeFormatter::create_ymdt(*locale.get(), options)
```

sffc commented:
I will also take feedback on the format functions.

I intend to add the following format functions:

1. `format` takes the same type as Rust `format` (matching calendar in GregorianDateTimeFormatter, any calendar in DateTimeFormatter)
2. `format_same_calendar` has the same behavior as Rust `format_same_calendar`
3. `format_iso` exists only in DateTimeFormatter

We don't have `format_iso` in Rust but I put it in C++ since we don't have overloading or traits.

Also, please notice that `format_iso` takes two positional arguments, the date and the time. I did not name the function `format_iso_datetime`. Instead, the required arguments come from the type. `DateTimeFormatter`'s format functions take a date and a time as positional arguments.

Manishearth reviewed (approved):
Seems like the right approach!

sffc commented:
Action on me:

1. Change the Dart tests to use the new FFI so that Robert can review what it looks like in Dart.
2. (ideally after Rob leaves feedback above, but before then if necessary) Remove the "old" FFI for DateTimeFormatter and migrate the tests to the new FFI

sffc commented on ffi/dart/test/icu_test.dart:
```diff
@@ -61,6 +61,25 @@ void main() {
+      dateFields: DateFields.mde,
+      length: DateTimeLength.short,
+      timePrecision: TimePrecision.minute,
+      zoneStyle: null,
+      alignment: null,
+      yearStyle: null,
```
Note: The generated code seems to make it required that I set these fields, even though their type is an `Option`.

sffc commented on ffi/dart/test/icu_test.dart:
```diff
@@ -61,6 +61,25 @@ void main() {
+      yearStyle: null,
+    );
+
+    expect(
+        DateTimeFormatter.fromBuilder(locale, builder)
+            .formatIso(zonedDateTime.date, zonedDateTime.time),
+        'Mi., 15.01., 14:32');
+
+    expect(
+        DateTimeFormatter.mdet(locale, DateTimeLength.short, TimePrecision.minute, DateTimeAlignment.auto)
+            .formatIso(zonedDateTime.date, zonedDateTime.time),
+        'Mi., 15.01., 14:32');
```
@robertbastian Here is the Dart code.

robertbastian commented on ffi/capi/src/neo_datetime.rs:
```diff
@@ -0,0 +1,505 @@
+    }
+
+    #[diplomat::rust_link(icu::datetime::builder::FieldSetBuilder, Enum)]
+    pub struct DateTimeFieldSetBuilder {
```
is this really a builder? it's useable without having to call `.build()` on it, it's more of an options bag. consider renaming to `DateTimeFieldSet`

robertbastian commented on ffi/capi/src/neo_datetime.rs:
```diff
@@ -0,0 +1,505 @@
+        #[cfg(feature = "compiled_data")]
+        pub fn create_dt(
+            locale: &Locale,
+            length: DateTimeLength,
+            time_precision: TimePrecision,
+            alignment: DateTimeAlignment,
```
should these be optional?

sffc replied:
This is exactly one of things this PR is intended to be a place to bikeshed. I posted several choices in https://github.com/unicode-org/icu4x/pull/6035#issuecomment-2614229124. Positional option arguments is another approach I didn't consider.

robertbastian replied:
The choices in the comment bunch together multiple axes: whether to use options bags or parameters, and whether the options are options. I don't think we should ever force a user to supply all arguments, so the three APIs we should discuss are single options bag, options bag per type, and option parameters.

sffc commented:
I changed it to optional positional arguments, and I removed the FFI for the field set builder. I will add it back when I add time zone formatting. https://github.com/unicode-org/icu4x/pull/6035/commits/aa85e53b5f22b4978af055b7335126fcaecd02ce

robertbastian reviewed (commented):
Happy with this API, please go ahead and add the others

sffc commented on tutorials/cpp/datetime.cpp:
```diff
@@ -24,11 +25,80 @@ int main() {
     std::unique_ptr<IsoDate> date = IsoDate::create(2022, 07, 11).ok().value();
     std::unique_ptr<Time> time = Time::create(13, 06, 42, 0).ok().value();
 
+    std::unique_ptr<DateTimeFormatter> fmt_dt = DateTimeFormatter::create_dt(*locale.get(), DateTimeLength::Medium, TimePrecision::Minute, std::optional<DateTimeAlignment>()).ok().value();
```
Yes; it is constructing an instance of type `std::optional<DateTimeAlignment>` with no arguments `()` and is therefore None. I didn't spend a _lot_ of time seeing if there was a more concise option.

sffc commented:
> Happy with this API, please go ahead and add the others

I'd like to land this PR as-is and follow up with at least two more PRs:

1. Migrate the rest of the non-zoned things (not controversial, just tedious)
2. Add the zoned formatter (needs design)

robertbastian commented:
Can you do `GregorianDateTimeFormatter` now?

sffc commented:
> Can you do `GregorianDateTimeFormatter` now?

ok. Done.

sffc commented:
Disclaimer: I have only implemented the `format_iso` function as of now. I may try pushing the other format signatures.
